src/troposphere_dns_certificate/certificate.py

```diff
@@ -65,43 +62,54 @@ def add_tags(arn, p):
 
 def get_zone_for(name, p):
     name = name.rstrip('.')
-    hosted_zones = {v['DomainName'].rstrip('.'): v['HostedZoneId'] for v in p['DomainValidationOptions']}
+    zones = {v['DomainName'].rstrip('.'): v['HostedZoneId'] for v in p['DomainValidationOptions']}
 
-    components = name.split('.')
+    parts = name.split('.')
 
-    while len(components):
-        if '.'.join(components) in hosted_zones:
-            return hosted_zones['.'.join(components)]
+    while len(parts):
+        if '.'.join(parts) in zones:
+            return zones['.'.join(parts)]
 
-        components = components[1:]
+        parts = parts[1:]
 
     raise RuntimeError('DomainValidationOptions' + ' missing for %s' % str(name))
 
 
-def validate(arn, p):
+def validate(e, p):
     if 'ValidationMethod' in p and p['ValidationMethod'] == 'DNS':
 
-        all_records_created = False
-        while not all_records_created:
-            all_records_created = True
+        done = False
+        while not done:
+            done = True
 
-            certificate = acm.describe_certificate(CertificateArn=arn)['Certificate']
-            l.info(certificate)
+            cert = acm.describe_certificate(CertificateArn=e['PhysicalResourceId'])['Certificate']
+            l.info(cert)
 
-            if certificate['Status'] != 'PENDING_VALIDATION':
+            if cert['Status'] != 'PENDING_VALIDATION':
                 return
 
-            for v in certificate['DomainValidationOptions']:
+            for v in cert['DomainValidationOptions']:
 
                 if 'ValidationStatus' not in v or 'ResourceRecord' not in v:
-                    all_records_created = False
+                    done = False
                     continue
 
                 if v['ValidationStatus'] == 'PENDING_VALIDATION':
-                    response = boto3.client('route53').change_resource_record_sets(
+                    c = {}
+                    if 'Route53RoleArn' in p:
+                        c = client('sts').assume_role(
+                            RoleArn=p['Route53RoleArn'],
+                            RoleSessionName='DNSCertificate'+e['LogicalResourceId'],
```

Comment:
The role session name has to be alpha-numeric (plus some special characters). Since the `LogicalResourceId` is strictly alpha-numeric, this cannot be a name with invalid characters.

The maximum length [is 64 characters](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html). Should we take a substring here to ensure this?

Reply:
Yeah, better truncate to be safe.

Reply:
I pushed an update that truncates the role session name.